File: src/detectors/apps.ts
import { dirname } from "node:path";
import type { Repo } from "../repo.js";
import type { Signal } from "../types.js";
import { FRAMEWORK_BY_DEPENDENCY } from "./framework.js";
import { gemfiles, nodeManifests, pythonManifestFiles } from "./manifest.js";

/**
 * How many separately deployable applications this repository holds.
 *
 * A verdict describes one system. When a repository holds several applications
 * the single answer is still useful, but it is an answer about the whole, and
 * the owner deserves to be told that rather than left to assume otherwise.
 *
 * An application root is a directory whose own manifest declares a web
 * framework. A directory with a manifest and no framework is a package, not a
 * deployment, so it does not count.
 */
export function detectApps(repo: Repo): Signal[] {
  const roots = new Map<string, string>();

  for (const [file, manifest] of nodeManifests(repo)) {
    const names = new Set<string>();
    for (const key of ["dependencies", "devDependencies"]) {
      const block = manifest[key];
      if (typeof block !== "object" || block === null) continue;
      for (const name of Object.keys(block)) names.add(name.toLowerCase());
    }
    noteRoot(roots, file, names);
  }

  for (const file of [...pythonManifestFiles(repo), ...gemfiles(repo)]) {
    const text = (repo.read(file) ?? "").toLowerCase();
    const names = new Set([...FRAMEWORK_BY_DEPENDENCY.keys()].filter((name) => text.includes(name)));
    noteRoot(roots, file, names);
  }

  const count = roots.size;
  if (count > 1) {
    return [
      {
        kind: "apps",
        values: ["several"],
        confidence: "high",
        metric: count,
        evidence: `${count} application roots: ${[...roots.values()].join(", ")}`,
      },
    ];
  }

  return [
    {
      kind: "apps",
      values: ["one"],
      confidence: count === 1 ? "high" : "low",
      metric: count,
      evidence:
        count === 1
          ? `one application root at ${[...roots.values()][0]}`
          : "no manifest declares a web framework, so there is no application root to count",
    },
  ];
}

function noteRoot(roots: Map<string, string>, file: string, names: Set<string>): void {
  for (const name of names) {
    if (!FRAMEWORK_BY_DEPENDENCY.has(name)) continue;
    const root = dirname(file);
    if (!roots.has(root)) roots.set(root, file);
    return;
  }
}

File: internal/scan/apps.go
package scan

import (
	"fmt"
	"path"
	"strings"
)

// DetectApps reports how many separately deployable applications this
// repository holds.
//
// A verdict describes one system. When a repository holds several applications
// the single answer is still useful, but it is an answer about the whole, and
// the owner deserves to be told that rather than left to assume otherwise.
//
// An application root is a directory whose own manifest declares a web
// framework. A directory with a manifest and no framework is a package, not a
// deployment, so it does not count.
func DetectApps(r *Repo) []Signal {
	roots := newAppRoots()

	for _, manifest := range NodeManifests(r) {
		names := map[string]bool{}
		for _, key := range []string{"dependencies", "devDependencies"} {
			block, ok := manifest.Fields[key].(map[string]any)
			if !ok {
				continue
			}
			for name := range block {
				names[strings.ToLower(name)] = true
			}
		}
		roots.note(manifest.Path, names)
	}

	// Python and Ruby manifests are read as text rather than parsed, so the
	// framework names are looked for in the file itself. A manifest that
	// mentions one is the application root whichever key it appeared under.
	for _, file := range append(PythonManifestFiles(r), Gemfiles(r)...) {
		text := strings.ToLower(r.Read(file))
		names := map[string]bool{}
		for name := range FrameworkByDependency {
			if strings.Contains(text, name) {
				names[name] = true
			}
		}
		roots.note(file, names)
	}

	count := len(roots.order)
	if count > 1 {
		signal := Found(FieldApps, High,
			fmt.Sprintf("%d application roots: %s", count, strings.Join(roots.manifests(), ", ")),
			"several")
		signal.Metric = count
		return []Signal{signal}
	}

	confidence := Low
	evidence := "no manifest declares a web framework, so there is no application root to count"
	if count == 1 {
		confidence = High
		evidence = "one application root at " + roots.manifests()[0]
	}
	signal := Found(FieldApps, confidence, evidence, "one")
	signal.Metric = count
	return []Signal{signal}
}

// appRoots collects one manifest per directory that declares a web framework.
// Insertion order is kept so the evidence names roots in the order they were
// found rather than in whatever order a map hands back.
type appRoots struct {
	byDirectory map[string]string
	order       []string
}

func newAppRoots() *appRoots {
	return &appRoots{byDirectory: map[string]string{}}
}

// note records the manifest's directory as an application root when any of the
// names it declares is a web framework.
func (a *appRoots) note(file string, names map[string]bool) {
	for name := range names {
		if _, framework := FrameworkByDependency[name]; !framework {
			continue
		}
		root := path.Dir(file)
		if _, seen := a.byDirectory[root]; !seen {
			a.byDirectory[root] = file
			a.order = append(a.order, root)
		}
		return
	}
}

func (a *appRoots) manifests() []string {
	files := make([]string, 0, len(a.order))
	for _, root := range a.order {
		files = append(files, a.byDirectory[root])
	}
	return files
}
